Python/flowgraph.c

```diff
@@ -640,42 +645,54 @@ _PyCfg_Stackdepth(basicblock *entryblock, int code_flags)
         return ERROR;
     }
 
+
+    int stackdepth = -1;
     int maxdepth = 0;
     basicblock **sp = stack;
-    if (code_flags & (CO_GENERATOR | CO_COROUTINE | CO_ASYNC_GENERATOR)) {
-        stackdepth_push(&sp, entryblock, 1);
```

Comment:
I'm confused about this. Why is it 1 if the RETURN_GENERATOR+POP_TOP add up to 0?

Reply:
I wonder if it's somehow related to the fact that stack_effect() returns 0 for RETURN_GENERATOR?

Reply:
I added some prints to the old and new versions, to see how the stackdepth calculation proceeds. So here is the output for  `def f(): yield 12`.

Old:

```
1 [0]  opcode = 75  depth = 1 effect = 0  new_depth = 1  maxdepth = 1
1 [1]  opcode = 1  depth = 1 effect = -1  new_depth = 0  maxdepth = 1
1 [2]  opcode = 257  depth = 0 effect = 0  new_depth = 0  maxdepth = 1
2 [2]  opcode = 257  depth = 0 effect = 2  target_depth = 2  maxdepth = 2
1 [3]  opcode = 151  depth = 0 effect = 0  new_depth = 0  maxdepth = 2
1 [4]  opcode = 100  depth = 0 effect = 1  new_depth = 1  maxdepth = 2
1 [5]  opcode = 150  depth = 1 effect = 0  new_depth = 1  maxdepth = 2
1 [6]  opcode = 151  depth = 1 effect = 0  new_depth = 1  maxdepth = 2
1 [7]  opcode = 1  depth = 1 effect = -1  new_depth = 0  maxdepth = 2
1 [8]  opcode = 121  depth = 0 effect = 0  new_depth = 0  maxdepth = 2
1 [0]  opcode = 173  depth = 2 effect = 0  new_depth = 2  maxdepth = 2
1 [1]  opcode = 119  depth = 2 effect = -1  new_depth = 1  maxdepth = 2
1 [0]  opcode = 151  depth = 0 effect = 0  new_depth = 0  maxdepth = 0
1 [1]  opcode = 100  depth = 0 effect = 1  new_depth = 1  maxdepth = 1
1 [2]  opcode = 24  depth = 1 effect = 0  new_depth = 1  maxdepth = 1
1 [3]  opcode = 90  depth = 1 effect = -1  new_depth = 0  maxdepth = 1
1 [4]  opcode = 121  depth = 0 effect = 0  new_depth = 0  maxdepth = 1
```

New:

```
1 [0]  opcode = 257  depth = 0 effect = 0  new_depth = 0  maxdepth = 0
2 [0]  opcode = 257  depth = 0 effect = 2  target_depth = 2  maxdepth = 2
1 [1]  opcode = 151  depth = 0 effect = 0  new_depth = 0  maxdepth = 2
1 [2]  opcode = 100  depth = 0 effect = 1  new_depth = 1  maxdepth = 2
1 [3]  opcode = 150  depth = 1 effect = 0  new_depth = 1  maxdepth = 2
1 [4]  opcode = 151  depth = 1 effect = 0  new_depth = 1  maxdepth = 2
1 [5]  opcode = 1  depth = 1 effect = -1  new_depth = 0  maxdepth = 2
1 [6]  opcode = 121  depth = 0 effect = 0  new_depth = 0  maxdepth = 2
1 [0]  opcode = 173  depth = 2 effect = 0  new_depth = 2  maxdepth = 2
1 [1]  opcode = 119  depth = 2 effect = -1  new_depth = 1  maxdepth = 2
1 [0]  opcode = 151  depth = 0 effect = 0  new_depth = 0  maxdepth = 0
1 [1]  opcode = 100  depth = 0 effect = 1  new_depth = 1  maxdepth = 1
1 [2]  opcode = 24  depth = 1 effect = 0  new_depth = 1  maxdepth = 1
1 [3]  opcode = 90  depth = 1 effect = -1  new_depth = 0  maxdepth = 1
1 [4]  opcode = 121  depth = 0 effect = 0  new_depth = 0  maxdepth = 1
```

So the old code looks at those two instructions which are now added after the stackdepth calculation. And indeed stackdepth(RETURN_GENERATOR) returns 0, so the additional 1 in the initialisation compensated for that. So the old code is a messy hack.

Reply:
I wondered the same thing! Thanks for tracking this down. Separately, is there any reason not to fix num_pushed metadata for `RETURN_GENERATOR`? I guess with this PR it now hardly matters what that metadata says for it...

Reply:
The problem is that the stack effect is derived from the bytecodes.c file, from the signature of the bytecode. If we add stuff to the signature the code generator will do other stuff with it.